### PR TITLE
Graphnode caching before execution

### PR DESCRIPTION
**Pull request by junmendoza**
### Purpose

1. The VM caches the graphnodes that it is about to execute in the current scope. (Graphnodes in the current language block or function) 
2. This avoids redundant calls GetGraphNodes in scope
3. This change is required to help implement macroblocks. When code is fragmented to macroblocks, it is important to cache only those graphnodes that belong to the current scope and macroblock. This makes it easier to implement and maintain macroblocks 

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files

### Reviewers

@ke-yu 

### FYIs

Ive disable one language test that is an edgecase which involves calling a function where the code contains syntax errors. It must be fixed such that the function should not be called at all.

### CODE REVIEW

File: src/Engine/ProtoCore/AssociativeGraph.cs [hunk 1]
```diff
@@ -16,6 +16,69 @@ public enum UpdateStatus
     public class Utils
     {
         /// <summary>
```
Comment (junmendoza) [inline src/Engine/ProtoCore/AssociativeGraph.cs]: Move functions from DependencyGraph to Utils.

File: src/Engine/ProtoCore/DSASM/Executive.cs [hunk 2]
```diff
@@ -108,6 +114,21 @@ public Executive(RuntimeCore runtimeCore, bool isFep = false)
         }
 
         /// <summary>
+        /// Cache the graphnodes in scope
+        /// </summary>
+        private void SetupGraphNodesInScope()
```
Comment (junmendoza) [inline src/Engine/ProtoCore/DSASM/Executive.cs]: This function caches the graphnodes in the scope that is about to execute

File: src/Engine/ProtoCore/AssociativeGraph.cs [hunk 3]
```diff
@@ -16,6 +16,69 @@ public enum UpdateStatus
...
+        /// Gets the dirty graphnode of the given pc
+        /// </summary>
+        /// <param name="pc"></param>
+        /// <param name="classIndex"></param>
+        /// <param name="procIndex"></param>
+        /// <returns></returns>
+        public static AssociativeGraph.GraphNode GetGraphNodeAtPC(int pc, List<AssociativeGraph.GraphNode> graphNodesInScope)
+        {
+            if (graphNodesInScope != null && graphNodesInScope.Count > 0)
```
Comment (ke-yu) [inline src/Engine/ProtoCore/AssociativeGraph.cs]: I'd suggest nullity test could be done by assertion. And `graphNodesInScope.Count > 0` is redundant.

File: src/Engine/ProtoCore/AssociativeGraph.cs [hunk 4]
```diff
@@ -16,6 +16,69 @@ public enum UpdateStatus
...
+        /// </summary>
+        /// <param name="pc"></param>
+        /// <param name="classIndex"></param>
+        /// <param name="procIndex"></param>
+        /// <returns></returns>
+        public static AssociativeGraph.GraphNode GetFirstDirtyGraphNodeFromPC(int pc, List<AssociativeGraph.GraphNode> graphNodesInScope)
+        {
+            if (graphNodesInScope != null && graphNodesInScope.Count > 0)
+            {
+                foreach (AssociativeGraph.GraphNode gnode in graphNodesInScope)
```
Comment (ke-yu) [inline src/Engine/ProtoCore/AssociativeGraph.cs]: Just for your reference, this function could be replaced by
`graphNodesInScope.FirstOrDefault(g => g.isActive && g.isDirty && g.updateBlock.startpc >= pc)`

Comment (ke-yu) [inline src/Engine/ProtoCore/AssociativeGraph.cs]: Similar for the previous function. :-)

File: src/Engine/ProtoCore/AssociativeGraph.cs (hunk 3 above)
Comment (junmendoza) [inline src/Engine/ProtoCore/AssociativeGraph.cs]: Done

File: src/Engine/ProtoCore/AssociativeGraph.cs (hunk 4 above)
Comment (junmendoza) [inline src/Engine/ProtoCore/AssociativeGraph.cs]: Done! thanks

**Pull request merged**